### ISSUE TITLE
Doesn't start on a Raspberry Pi 4

### ISSUE THREAD

**Issue by JiffB** [vkmark/vkmark#27]
Hi,

when launching, I just have one line of error ; here is the output of : vkmark -d : 
```
pi@rpi:~ $ vkmark -d
Debug: WindowSystemLoader: Looking in /usr/local/lib/arm-linux-gnueabihf/vkmark for window system plugins
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... ok
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... ok
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... succeeded with priority 127
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... succeeded with priority 0
Debug: WindowSystemLoader: Selected window system plugin /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so (best match)
Debug: XcbNativeSystem: Using root visual 0x21 for window
Error: vk::Instance::enumeratePhysicalDevices: ErrorInitializationFailed
```
Do you have an idea about what's going wrong?

**Comment by Bukkk** [comment]
Hello

This error is not vkmark specific. It is defined in vulkan.hpp (vulkan c++ wrapper) in line 82661 and later thrown in line 13785 after error code to message string conversion. Vkmark just catches all uncaught exceptions in main function and adds "Error: " at the beginning.

Are you sure you have your vulkan drivers SUCCESSFULLY installed? (similar issue?: https://github.com/vkmark/vkmark/issues/16)
Have you tried running any other vulkan application?
Anything c++ written using vulkan.hpp?
Does raspberry pi 4 drivers fully and reliably support vulkan?

I found only this:
https://www.raspberrypi.org/blog/vulkan-update-now-with-added-source-code/

Maybe it is worth to open an issue on their project as well...
Or maybe their vulcan implementation needs some unique extension/s to work (other than "VK_KHR_xcb_surface")???

**Comment by JiffB** [comment]
@Bukkk Hi,

> Are you sure you have your vulkan drivers SUCCESSFULLY installed? (similar issue?: #16)

Normally, yes, all packages about vulkan are installed

> Have you tried running any other vulkan application?

Just find a pkg with examples (gears & triangle), unfortunately, both are failing with almost the same lines : 
```
err:vkd3d_select_physical_device: Failed to enumerate physical devices, vr -3.
fixme:hresult_from_vk_result: Unhandled VkResult -3.
gears-vkd3d: demos/gears.c:331: cxg_load_pipeline: Assertion `SUCCEEDED(hr)' failed.
Abandon

DEMO 2 FAIL ≠ triangle-vkd3d: demos/triangle.c:187: cxg_load_pipeline: Assertion `SUCCEEDED(hr)' failed.
```
Hmm, that doesn't talk to me.

I read the link you gave me, it seems very alpha at this time :/
I'll try to compile it from their git repo and will keep you aware about the result.

Thanks.

**Comment by JiffB** [comment]
Raahhh, doesn't work :(

Here is the result with the compiled driver (recipe from: https://blogs.igalia.com/apinheiro/2020/06/v3dv-quick-guide-to-build-and-run-some-demos/) : 
```
$ export VK_ICD_FILENAMES=/home/pi/local-install/share/vulkan/icd.d/broadcom_icd.arm.json
$ gear-vkd3d
vk: error: v3dv_GetPhysicalDeviceProperties2: ignored VkStructureType 1000212000:VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PCI_BUS_INFO_PROPERTIES_EXT

WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
err:vkd3d_check_extensions: Required device extension "VK_KHR_maintenance1" is not supported.
err:vkd3d_check_extensions: Required device extension "VK_KHR_shader_draw_parameters" is not supported.
err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -7.
gears-vkd3d: demos/gears.c:331: cxg_load_pipeline: Assertion `SUCCEEDED(hr)' failed.
```
Examples from KhronosGroup compile, but the final link fails :((
(The link he gives goes to a git whose owner explains he'll now participate to this other (official) git : https://github.com/KhronosGroup/Vulkan-Samples)

I guess I'll have to wait until a new stable version is introduced into Raspberry OS.

**Comment by Bukkk** [comment]
@JiffB Hi again,

I read the guide how to build drivers you sent. In commentary section i have found:
![Screenshot_20200707-061702](https://user-images.githubusercontent.com/44477673/86709286-14457380-c01a-11ea-9714-8ad6d6b05376.jpg)

**Comment by JiffB** [comment]
I totally missed thid one :/

That was my conclusion anyway and at this time I have another SD card w/ Raspberry OS 32 bit that just finished updated and will stay a pure 32 bit.
I'm now going to bed and will re-test vulkan new compilation + vkmark this evening or during the night.

Thanks for this information.

**Comment by infapi00** [comment]
Hi, by chance I found that issue. FWIW, around the time that comment was written there, I tested vkmark. In fact it was recommended by some of the people adding comments on that blog post. 

I remember that I was able to get vkmark running, but not with the default command line options. FWIW, that was needed because our driver didn't implement something specific. So it is really likely that we can just close this issue. 

I will try to get it running again today/tomorrow, and write here what was the issue.

**Comment by infapi00** [comment]
Well, it was fast. I didn't need to do anything special. It worked without doing anything special, and running just how the README.md says:
`build/src/vkmark --winsys-dir=build/src --data-dir=data`

I algo got this score:

>                                    vkmark Score: 348

I lack the reference to know if that is good or bad ;)

![vkmark1](https://user-images.githubusercontent.com/598414/97592254-bab9e800-1a00-11eb-8040-e2bfe25e76c8.jpg)

**Comment by infapi00** [comment]
@Bukkk so perhaps you can just close this issue. Im not sure what was the problem on July for @JiffB , but then the driver was more wip then.

@JiffB btw, if you want to test it again, you can go directly to Mesa master now. Unfourtunately it is not still available as a Raspbian package.

**Comment by JiffB** [comment]
> @JiffB btw, if you want to test it again, you can go directly to Mesa master now. Unfourtunately it is not still available as a Raspbian package.

Ok, but I spare it for another day|night, because this gonna be a looong month :/

**Comment by Bukkk** [comment]
Hi @infapi00 

I cannot believe you found this issue. Thank you for sharing your knowledge with us. It is nice to hear that it all works properly now and is merged into mesa :)

I would really like to close it, but i am not a collaborator, only a fork owner waiting for pull request review. We have to wait for @afrantzis .

> I lack the reference to know if that is good or bad ;)

FWIW if you are curious:
mesa/DRI3 skylake desktop iGPU achieves over 1500
mesa/amdgpu rx5700xt -- about 24000 (on wayland. on Xserver it is a little bit slower)
mesa/nouveau nvidia gt520 (series which is fairly dececnt in case of performance on nouveau) -- does 120 BUT on glmark2 (basically the same as vkmark but on opengl). No vulkan implementatnion for noveau as far as i know...

I was curious how well does it run on raspberry pi 4, but i didn't bother to ask @JiffB :)
And i am actually surprised.

Thank you once again

**Comment by JiffB** [comment]
@infapi00 > you can go directly to Mesa master now
I want, but what is it and where is it?

@Bukkk 
Ok, I restarted from a fresh clone but it still doesn't work, however messages have changed : 
Command line dry : 
```
# vkmark 
error: XDG_RUNTIME_DIR not set in the environment.
Error: Failed to find usable window system, try using --winsys-dir
```
Command line from git clone : 
```
# ./build/src/vkmark --winsys-dir=./build/src --data-dir=./data
error: XDG_RUNTIME_DIR not set in the environment.
Error: Failed to find usable window system, try using --winsys-dir
```
Debug : 
```
# vkmark -d
Debug: WindowSystemLoader: Looking in /usr/local/lib/arm-linux-gnueabihf/vkmark for window system plugins
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... ok
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... ok
error: XDG_RUNTIME_DIR not set in the environment.
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... succeeded with priority 0
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... succeeded with priority 0
Error: Failed to find usable window system, try using --winsys-dir
```
So it still breaks, sniff.

**Comment by Bukkk** [comment]
@JiffB 
Lines 2 and 3 means that vkmark found WSplugins, and can access functions inside those plugins.

I don't know where does the line 4 come from (probably you have gnome derrived desktop enviroment) -- this is not written here by vkmark (i am pretty sure)

The lines 5 and 6 priority 0 means that although vkmark can access the plugin, the plugin cannot open and close connection to display server without errors (probe a window system). To be precise, in case of Xserver, it means that Xserver returns error code when vkmark tries to open beforementioned connection.

Last line is just default exception thrown when vkmark cannot set up window system no matter the reason. Yeah, it is confusing :/

To sum it up, i think there is a problem with your configuration. I would start with googling for XDG_RUNTIME_DIR and trying solves from ubuntu/arch/manjaro forums, or just fresh system install.
something like this: https://askubuntu.com/questions/456689/error-xdg-runtime-dir-not-set-in-the-environment-when-attempting-to-run-naut

**Comment by Bukkk** [comment]
if it would help you my XDG_RUNTIME_DIR points to: /run/user/1000

**Comment by JiffB** [comment]
@Bukkk > I don't know where does the line 4 come from
My WM is XFCE.
> the plugin cannot open and close connection to display server without errors

Oh, ok my bad, I was running it as root :/
But it doesn't change much,
without an export of XDG_RUNTIME_DIR : 
```
$ vkmark 
Error: vk::Instance::enumeratePhysicalDevices: ErrorInitializationFailed
```
with :   (it is ok, as pi is effectively UID:GID 1000:1000)
```
$ export XDG_RUNTIME_DIR=/run/user/1000
$ vkmark 
Error: vk::Instance::enumeratePhysicalDevices: ErrorInitializationFailed
```
ze same :/

I see a window in a flash but that's all.

Debug : 
```
$ vkmark -d
Debug: WindowSystemLoader: Looking in /usr/local/lib/arm-linux-gnueabihf/vkmark for window system plugins
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... ok
Debug: WindowSystemLoader: Loading options from /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... ok
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/wayland.so... succeeded with priority 0
Debug: WindowSystemLoader: Probing /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so... succeeded with priority 127
Debug: WindowSystemLoader: Selected window system plugin /usr/local/lib/arm-linux-gnueabihf/vkmark/xcb.so (best match)
Debug: XcbNativeSystem: Using root visual 0x21 for window
Error: vk::Instance::enumeratePhysicalDevices: ErrorInitializationFailed
```
which brings back to my original post, as all lines are equals.

I forgot : using the line from your URL, I tried it, it asked my password (for the use of 'env' as pi or root) and failed as unauthorized - removing the leading 'pkexec', it failed as above.

**Comment by JiffB** [comment]
May be it is something related to the Pi 4 graphical structure, as redshift has no effect at all on it (red-ish the monitor depending on your LAT/LON and local time to avoid too much blue at night).

**Comment by Bukkk** [comment]
@Jifff
Have you installed mesa from source?? As @infapi00 wrote
> Unfourtunately it is not still available as a Raspbian package.

> @infapi00 > you can go directly to Mesa master now
I want, but what is it and where is it?

https://github.com/mesa3d/mesa

https://github.com/mesa3d/mesa/commits?author=infapi00

It should be well documented, but i have no clue how to install it. You have to ask @infapi00  :/

**Comment by JiffB** [comment]
@Bukkk > Have you installed mesa from source??
Nope, I have a lot of testing coming on this SBC and not very much free time, I also can't risk to have to reinstall it from scratch (many compilations on). I'll wait until a new mesa package comes up - thanks to the both of you for your time.

**Comment by infapi00** [comment]
> https://github.com/mesa3d/mesa

This is not the official repository, seems to be a mirror. This is the official repository:
https://gitlab.freedesktop.org/mesa/

> It should be well documented, but i have no clue how to install it. You have to ask @infapi00 :/

The instructions are basically the same that I put here:
https://blogs.igalia.com/apinheiro/2020/06/v3dv-quick-guide-to-build-and-run-some-demos/

But replacing the repository there with mesa master.

**Comment by Bukkk** [comment]
Hi again

I bought Raspberry Pi 4 recently. Now it works out of the box on 64bit Manjaro, after you install vulkan-broadcom package.

`Debug: WindowSystemLoader: Looking in /usr/lib/vkmark for window system plugins
Debug: WindowSystemLoader: Loading options from /usr/lib/vkmark/wayland.so... ok
Debug: WindowSystemLoader: Loading options from /usr/lib/vkmark/xcb.so... ok
Debug: WindowSystemLoader: Probing /usr/lib/vkmark/wayland.so... succeeded with priority 255
Debug: WindowSystemLoader: Probing /usr/lib/vkmark/xcb.so... succeeded with priority 127
Debug: WindowSystemLoader: Selected window system plugin /usr/lib/vkmark/wayland.so (best match)
Debug: Trying to use first supported device.
Debug: First supported device choosen!
Debug: VulkanState: Using queue family index 0 for WSI operations
Debug: VulkanState: Using queue family index 0 for rendering
Debug: SwapchainWindowSystem: Available surface format B8G8R8A8Srgb
Debug: SwapchainWindowSystem: Available surface format B8G8R8A8Unorm
Debug: SwapchainWindowSystem: Selected swapchain format B8G8R8A8Srgb
Debug: SwapchainWindowSystem: Swapchain contains 4 images
Info: =======================================================
Info:     vkmark 2017.08
Info: =======================================================
Info:     Vendor ID:      0x14E4
Info:     Device ID:      0x2A
Info:     Device Name:    V3D 4.2
Info:     Driver Version: 88084486
Info:     Device UUID:    6b6827bf682010e31e0b5b1b7d947310
Info: =======================================================
Info: [vertex] device-local=true: FPS: 995 FrameTime: 1.005 ms
Info: [vertex] device-local=false: FPS: 1145 FrameTime: 0.873 ms
Info: [texture] anisotropy=0: FPS: 1075 FrameTime: 0.930 ms
Info: [texture] anisotropy=16: FPS: 1018 FrameTime: 0.982 ms
Info: [shading] shading=gouraud: FPS: 926 FrameTime: 1.080 ms
Info: [shading] shading=blinn-phong-inf: FPS: 774 FrameTime: 1.292 ms
Info: [shading] shading=phong: FPS: 622 FrameTime: 1.608 ms
Info: [shading] shading=cel: FPS: 638 FrameTime: 1.567 ms
Info: [effect2d] kernel=edge: FPS: 410 FrameTime: 2.439 ms
Info: [effect2d] kernel=blur: FPS: 146 FrameTime: 6.849 ms
Info: [desktop] <default>: FPS: 159 FrameTime: 6.289 ms
Info: [cube] <default>: FPS: 1410 FrameTime: 0.709 ms
Info: [clear] <default>: FPS: 1014 FrameTime: 0.986 ms
Info: =======================================================
Info:                                    vkmark Score: 794
Info: =======================================================
`
